kafkaesk/retry.py

```diff
@@ -0,0 +1,157 @@
+from typing import TYPE_CHECKING
+
+import datetime
+
+if TYPE_CHECKING:
+    from .app import Application
+    from .app import Subscription
+
+
+class RetryInfo(BaseModel):
+    retry_count: int
+    retry_delay: int = 0
+    retry_timestamp: datetime.datetime
+    publish_timestamp: datetime.datetime
+    publish_topic: str
+    error: str
+
+
+class RetryMessage(BaseModel):
+    retry_info: RetryInfo
+    original_record: ConsumerRecord
+
+    class Config:
+        arbitrary_types_allowed = True
+
+
+class FailureInfo(BaseModel):
+    retry_count: int
+    failure_timestamp: datetime.datetime
+    publish_topic: str
+    error: str
+
+
+class FailureMessage(BaseModel):
+    failure_info: FailureInfo
+    original_record: ConsumerRecord
+
+    class Config:
+        arbitrary_types_allowed = True
+
+
+class RetryPolicy(ABC):
+    def __init__(self) -> None:
+        self._initialized = False
+
+    async def initialize(self, app: "Application", subscription: "Subscription") -> None:
+
+        self._app = app
+        self._subscription = subscription
+
+        self._initialized = True
+
+    async def finalize(self) -> None:
+        self._initialized = False
+
+    async def __call__(self, record: ConsumerRecord, error: Exception) -> None:
+        if self._initialized is not True:
+            raise RuntimeError("RetryPolicy is not initialized")
+
+        if self._should_retry(record, error):
+            return await self._handle_retry(record, error)
+        else:
+            return await self._handle_failure(record, error)
+
+    def _should_retry(self, record: ConsumerRecord, error: Exception) -> bool:
+        # We can not recover from formatting problems with the message.  Do not attempt to retry.
+        if isinstance(error, UnhandledMessage):
+            return False
+
+        return self.should_retry(record, error)
+
+    async def _handle_retry(self, record: ConsumerRecord, error: Exception) -> None:
+        MESSAGE_REQUEUED.labels(
+            stream_id=record.topic,
+            partition=record.partition,
+            error=error.__class__.__name__,
+            group_id=self._subscription.group,
+        ).inc()
+
+        await self.handle_retry(record, error)
+
+    async def _handle_failure(self, record: ConsumerRecord, error: Exception) -> None:
+        MESSAGE_FAILED.labels(
+            stream_id=record.topic,
+            partition=record.partition,
+            error=error.__class__.__name__,
+            group_id=self._subscription.group,
+        ).inc()
+
+        try:
+            await self.handle_failure(record, error)
+        except UnhandledMessage:
+            # Here we swallow any errors related to input format to mantain existing functionality
+            pass
+
+    def should_retry(self, record: ConsumerRecord, error: Exception) -> bool:
+        raise NotImplementedError
+
+    async def handle_retry(self, record: ConsumerRecord, error: Exception) -> None:
+        raise NotImplementedError
```

Comment:
So is there a type that will ever immediately retry here? I think all our types of retries will be something along the lines of forwarding to another topic right? Maybe we don't need the `should_retry`, `handle_retry`, etc interfaces.

Maybe you only need the policy to implement `__call__` for how they want to handle the message?

I could be missing something though.

Reply:
You are right, we won't ever be retrying immediately and will allways be writing to a queue.  I think this could be re-named to `should_requeue` rather than `should_retry`.

My thoughts behind this layout of `hooks`, for lack of a better term, is to enable metrics, and other common information, in one place rather than having those calls spread across every policy implementation.

Reply:
gotcha, that makes sense for metrics, etc I think then.

Reply:
I've renamed `should_retry` to `should_requeue` as it's clearer as to what is happening.